test/testsyncmove.cpp

```diff
@@ -0,0 +1,414 @@
+        // fileid as existing ones
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+
+        int nGET = 0;
+        fakeFolder.setServerOverride([&](QNetworkAccessManager::Operation op, const QNetworkRequest &) {
+            if (op == QNetworkAccessManager::GetOperation)
+                ++nGET;
+            return nullptr;
+        });
+
+        // Try a remote file move
+        remote.rename("A/a1", "A/W/a1m");
+        remote.rename("O/A/a1", "O/A/W/a1m");
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+        QCOMPARE(nGET, 0);
+
+        // And a remote directory move
+        remote.rename("A/W", "A/Q/W");
+        remote.rename("O/A/W", "O/A/Q/W");
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+        QCOMPARE(nGET, 0);
+
+        // Partial file removal (in practice, A/a2 may be moved to O/a2, but we don't care)
+        remote.rename("O/A/a2", "O/a2");
+        remote.remove("A/a2");
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+        QCOMPARE(nGET, 0);
+
+        // Local change plus remote move at the same time
+        fakeFolder.localModifier().appendByte("O/a2");
+        remote.rename("O/a2", "O/a3");
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+        QCOMPARE(nGET, 1);
+    }
+
+    void testMovePropagation()
+    {
+        FakeFolder fakeFolder{ FileInfo::A12_B12_C12_S12() };
+        auto &local = fakeFolder.localModifier();
+        auto &remote = fakeFolder.remoteModifier();
+
+        int nMOVE = 0;
+        fakeFolder.setServerOverride([&](QNetworkAccessManager::Operation, const QNetworkRequest &req) {
+            if (req.attribute(QNetworkRequest::CustomVerbAttribute) == "MOVE")
+                ++nMOVE;
+            return nullptr;
+        });
+
+        // Move
+        {
+            local.rename("A/a1", "A/a1m");
+            remote.rename("B/b1", "B/b1m");
+            QSignalSpy completeSpy(&fakeFolder.syncEngine(), SIGNAL(itemCompleted(const SyncFileItemPtr &)));
+            QVERIFY(fakeFolder.syncOnce());
+            QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+            QCOMPARE(nMOVE, 1);
+            QVERIFY(itemSuccessfulMove(completeSpy, "A/a1m"));
+            QVERIFY(itemSuccessfulMove(completeSpy, "B/b1m"));
+        }
+
+        // Touch+Move on same side
+        local.rename("A/a2", "A/a2m");
+        local.setContents("A/a2m", 'A');
+        remote.rename("B/b2", "B/b2m");
+        remote.setContents("B/b2m", 'A');
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+        QCOMPARE(nMOVE, 1);
```

Comment:
maybe also compute the amount of GET/PUT/DELETE  to make sure that there as not too many.